src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -783,6 +637,78 @@ private Compilation CreateEmptyCompilation()
                 return finalState.HasSuccessfullyLoaded;
             }
 
+            public ICompilationTracker FreezePartialState(CancellationToken cancellationToken)
+            {
+                var state = this.ReadState();
+
+                // If we're finalized and already frozen, we can just use ourselves.
+                if (state is FinalCompilationTrackerState { IsFrozen: true } finalState)
+                    return this;
+
+                var projectState = state switch
+                {
+                    // If we don't have an existing state, then transition to a project state without any data.
+                    null => this.ProjectState.RemoveAllDocuments(),
+                    FinalCompilationTrackerState => this.ProjectState,
+                    // If we have an in progress state with no steps, then we're just at the current project state.
+                    InProgressState { PendingTranslationSteps: [] } => this.ProjectState,
+                    // Otherwise, reset us to whatever state the InProgressState had currently transitioned to.
+                    InProgressState inProgressState => inProgressState.PendingTranslationSteps.First().oldState,
+                    _ => throw ExceptionUtilities.UnexpectedValue(state.GetType()),
+                };
+
+                var frozenState = GetFrozenCompilationState(state);
+                Contract.ThrowIfFalse(frozenState.IsFrozen);
+                return new CompilationTracker(projectState, frozenState, this.SkeletonReferenceCache.Clone());
+
+                CompilationTrackerState GetFrozenCompilationState(CompilationTrackerState? state)
+                {
+                    if (state is FinalCompilationTrackerState finalState)
+                    {
+                        // Checked by caller.
+                        Contract.ThrowIfTrue(finalState.IsFrozen);
+                        // If we're already finalized then just return what we have, and with the frozen bit flipped so
+                        // that any future forks keep things frozen.
+                        return finalState.WithIsFrozen();
+                    }
+                    else if (state is null)
+                    {
+                        // We have no data at all. Create a frozen empty project/compilation to represent this state.
+                        var inProgressProject = this.ProjectState.RemoveAllDocuments();
```

Comment:
Didn't we do this in the caller? Why do we have two switches here? It seems like it'd be easier to just have this method return the tuple of both states at once.

Reply:
indeed.